cpp/src/arrow/flight/sql/odbc/tests/connection_test.cc

```diff
@@ -785,9 +785,123 @@ TEST(SQLGetDiagRec, TestSQLGetDiagRecForConnectFailure) {
 
   EXPECT_TRUE(ret == SQL_ERROR);
 
-  if (ret != SQL_SUCCESS) {
-    std::cerr << GetOdbcErrorMessage(SQL_HANDLE_DBC, conn) << std::endl;
-  }
+  // Retrieve all supported header level and record level data
+  SQLSMALLINT HEADER_LEVEL = 0;
+  SQLSMALLINT RECORD_1 = 1;
+
+  //
+  // SQL_DIAG_NUMBER
+  //
+  SQLINTEGER diag_number;
+  SQLSMALLINT diag_number_length;
+
+  ret = SQLGetDiagField(SQL_HANDLE_DBC, conn, HEADER_LEVEL, SQL_DIAG_NUMBER, &diag_number,
+                        sizeof(SQLINTEGER), &diag_number_length);
+
+  EXPECT_TRUE(ret == SQL_SUCCESS);
+
+  EXPECT_TRUE(diag_number == 1);
+
+  // SQL_DIAG_SERVER_NAME
+  //
+  SQLWCHAR server_name[ODBC_BUFFER_SIZE];
+  SQLSMALLINT server_name_length;
+
+  ret = SQLGetDiagField(SQL_HANDLE_DBC, conn, RECORD_1, SQL_DIAG_SERVER_NAME, server_name,
+                        ODBC_BUFFER_SIZE, &server_name_length);
+
+  EXPECT_TRUE(ret == SQL_SUCCESS);
+
+  // SQL_DIAG_MESSAGE_TEXT
+  //
+  SQLWCHAR message_text[ODBC_BUFFER_SIZE];
+  SQLSMALLINT message_text_length;
+
+  ret = SQLGetDiagField(SQL_HANDLE_DBC, conn, RECORD_1, SQL_DIAG_MESSAGE_TEXT,
+                        message_text, ODBC_BUFFER_SIZE, &message_text_length);
+
+  EXPECT_TRUE(ret == SQL_SUCCESS);
+
+  EXPECT_TRUE(message_text_length > 100);
+
+  // SQL_DIAG_NATIVE
+  //
+  SQLINTEGER diag_native;
+  SQLSMALLINT diag_native_length;
+
+  ret = SQLGetDiagField(SQL_HANDLE_DBC, conn, RECORD_1, SQL_DIAG_NATIVE, &diag_native,
+                        sizeof(diag_native), &diag_native_length);
+
+  EXPECT_TRUE(ret == SQL_SUCCESS);
+
+  EXPECT_TRUE(diag_native == 200);
+
+  // SQL_DIAG_SQLSTATE
+  //
+  const SQLSMALLINT sql_state_size = 6;
+  SQLWCHAR sql_state[sql_state_size];
+  SQLSMALLINT sql_state_length;
+  ret = SQLGetDiagField(SQL_HANDLE_DBC, conn, RECORD_1, SQL_DIAG_SQLSTATE, sql_state,
+                        sql_state_size * 2, &sql_state_length);
```

Comment:
```suggestion
  ret = SQLGetDiagField(SQL_HANDLE_DBC, conn, RECORD_1, SQL_DIAG_SQLSTATE, sql_state,
                        sql_state_size * driver::odbcabstraction::GetSqlWCharSize(), &sql_state_length);
```
we can use `GetSqlWCharSize` to be more safe, since the size of wchar is not guaranteed to be 2 in non-windows platforms. I think we should be able to use it directly because we are including `encoding_utils.h` in the `odbc_test_suite.h` header file, but pls double-check